### PR TITLE
InstructorFeedbackSessionActions : Session.isClosed() returns wrong value for Private sessions with No specified End Time #8406

### LINKED ISSUE #8406: InstructorFeedbackSessionActions : Session.isClosed() returns wrong value for Private sessions with No specified End Time
**Environment:**
`master` branch

While working on #8374 I changed the condition for `isAllowedToSubmit` in `InstructorFeedbackSessionActions.java` to disable `submit` button for Closed sessions, when I found out that in case of the `Private` sessions with no set `End Time` (Like the one given below), the value of the method `isClosed()` returns `true`, where if we compare the current time with `null` (which is what the `isClosed()` method does  when the end time is not specified) in a statement like
`return now.after(null)` or `return now.before(null)` , it should always returns false.<br>
screenshot 342

**Steps to reproduce:** To be sure of this (supposed) bug, I followed the following steps.
- Login as `Insturctor`
- Create a `Private` session without specifying a `Submission closing time`
- Insert these lines (To get the contradicting results) in the file `InstructorFeedbackSessionActions.java` within the `InstructorFeedbackSessionActions` constructor, **after creating one/multiple private sessions with no end time specified** 

```java
System.out.println("Name: " + session.getSessionName()+"\tEndTime:" + session.getEndTimeString()
        + "\tisClosed: " + session.isClosed());
double timeZone = session.getTimeZone();
Calendar now = TimeHelper.now(timeZone);
System.out.println(session.getSessionName()+": isAfterNull: " +now.after(null));
```
- After saving and building the project go back to the `Instructor Home page` and refresh it.
- Check out the console for the print statements and find the sessions you created as private and check the output. (Check the end time which should be `-`, which indicates that the end time is `null`)

**Expected behaviour:**
The value of `now.after(null)` in the second println statement and the value of `isClosed()` should match. As both of them are doing the same thing.

**Actual behaviour:** This is the output from my console, for the private sessions I created.

```
Name: Private	EndTime:-	isClosed: true
Private: isAfterNull: false
Name: Private yet to start	EndTime:-	isClosed: true
Private yet to start: isAfterNull: false
Name: Private with no EndTime	EndTime:-	isClosed: true
Private with no EndTime: isAfterNull: false

```
Additionally while I was working on #8374 I modified the condition for `isAllowedToSubmit` to this : 
` this.isAllowedToSubmit = (session.isVisible() || session.isPrivateSession()) && !session.isClosed() && shouldEnableSubmitLink;
`
which disabled the  `submit` button for all the private sessions without a specified end time, even if the session has already started ( current time is after `startTime`)
**[Screenshot]**
screenshot 341

### PR DESCRIPTION
**Pull request by sukanta-27**
Fixes #8406 

**Outline of Solution:** As @tran-tien-dat confirmed that this is a legit bug in https://github.com/TEAMMATES/teammates/issues/8406#issuecomment-363461359, and that this bug is localized and only occurs in a particular case where `endTime == null` (particularly for `private` sessions), so I Modified the `isClosed()` method in `FeedbackSessionAttributes.java` to return `false` if `endTime == null`, as if end time is set to be `null`, the value of `now.after(end)` should always return false.

### CODE REVIEW

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java [hunk 1]
```diff
@@ -283,9 +283,13 @@ public boolean isClosedWithinPastHour() {
     }
 
     /**
-     * Returns {@code true} if it is after the closing time of this feedback session; {@code false} if not.
+     * Returns {@code true} if it is after the closing time of this feedback session; {@code false} if not
+     * or if the closing time is {@code null}, specifically in case of private sessions.
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: No need for ~~, specifically in case of private sessions.~~

**Comment by sukanta-27** [comment]
@wkurniawan07 Just checked. And as weird as it may sound, `startTime` and `endTime` both can be `null` for `Private` sessions. Looking into this.
screenshot 350

**Comment by damithc** [comment]
Actually, private sessions are rarely used. I'm not opposed to getting rid of them altogether.

**Comment by sukanta-27** [comment]
>I'm not opposed to getting rid of them altogether

@damithc Sir, shouldn't this be another issue completely. If you and @wkurniawan07 decide to remove the `Private` sessions altogether, I can open a new issue and would be more than happy to work on it.

**Comment by damithc** [comment]
> @damithc Sir, shouldn't this be another issue completely.

Yes, it should be a separate issue.  We should find how many private sessions exist already. If the number is not very low, not only removing it will be harder, but also there is evidence that it is a useful feature. I suspect the number will be low, but need to confirm first. @wkurniawan07 what do you think?

**Comment by sukanta-27** [comment]
**Outline of Solution:**
`isOpened()`: returns `false` if `startTime == null` as the session will never start and `true` if `endTime == null` as the session will never end. Will return `false` if start and end time are both `null`.

`isClosed()`:  returns `false` if `endTime == null` as the session will never end.

`isInGracePeriod():` returns `false` if `endTime == null` as the session will never end, and thus will never likely go into grace period.

`isWaitingToOpen()`: returns `true` if `startTime == null` as the session will never start, and thus by default will always be in waiting period.

**Comment by sukanta-27** [comment]
@wkurniawan07 Ready for a review.

**Comment by sukanta-27** [comment]
`isClosedAfter(int hours)`:  will always return false if `startTime == null` as in this case the session will not even start and if `endTime == null` because in this case the session will never end.

`isClosingWithinTimeLimit`: follows the same logic as `isClosedAfter`method mentioned above.
Like `isClosedWithinPastHour()`, both of these methods do not directly use `dateToCalendar`, but the use `setTime()` and `getTime()` in these cases will throw `NullPointerException`.

**Comment by wkurniawan07** [comment]
@whipermr5 I think I have to pass this review to you since you're working on this area right now.

**Comment by whipermr5** [comment]
Oops, I realised I actually fixed this issue back when I was developing #8094 (specifically, [this](https://github.com/TEAMMATES/teammates/pull/8094/commits/3939b009e6bdf305ad02a15470afd1811dbaa997) commit). However I think this PR can remain open to update the documentation and handle null for the first three methods (even though they're never used). Can you resolve the conflicts before we continue?

**Comment by sukanta-27** [comment]
@whipermr5 Yes, I will resolve the conflicts.
By the way, just to clarify a confusion I have, in the commit you mentioned, in the method `isWaitingToOpen()`:
```diff
     public boolean isWaitingToOpen() {
+        if (startTimeUtc == null) {
+            return false;
+        }
+
         Date now = new Date();
 
         return now.before(startTimeUtc);
```

shouldn't this return `true`, as (In my commit), I thought if `startTimeUtc == null`, then the session will never open, doesn't that mean that technically the session is waiting to open?

**Comment by whipermr5** [comment]
I think my reasoning was that waiting to open implies that it will one day be open.

Anyway, I checked all usages of the method. In practice, a null start time means that the session is private to the instructor who created it. The usages in `AdminSessionsPageAction` and `AdminSessionsPageData` are only called for ongoing sessions, not private sessions. The usages in `StudentFeedbackSessionActions` and `StudentHomePageData` are not applicable as private sessions are not visible to students. In `PageData`, the method is only called for non-private sessions. The only real difference this makes is In `FeedbackSessionPublishButton`.

Taking into account that private sessions are never published, `isWaitingToOpen` should indeed return true so that the publish button displayed in the results page of private sessions will always be disabled:
https://github.com/TEAMMATES/teammates/blob/a6f487d8386f7649578f9ed23b6971f17aa22017/src/main/java/teammates/ui/template/FeedbackSessionPublishButton.java#L27-L44

However this should be fixed in the logic of `FeedbackSessionPublishButton` instead of depending on the result of `isWaitingToOpen`. Ideally this method should not be called at all for private sessions.

**Comment by sukanta-27** [comment]
@whipermr5 okay. 2 more queries. 
- about `isOpened()` method That I mentioned in https://github.com/TEAMMATES/teammates/pull/8408#issuecomment-364449199
- and should we fix the logic in `FeedbackSessionPublishButton` in this PR, or should I create a new issue for it?

**Comment by whipermr5** [comment]
> about isOpened() method That I mentioned in #8408 (comment)

I don't think there's a point discussing what the right return values are when these methods were never meant to be called for private sessions in the first place. The important thing is how these methods are currently being used.

> and should we fix the logic in FeedbackSessionPublishButton in this PR, or should I create a new issue for it?

I think we can. This issue/PR can be about fixing the places that misuse these methods.

**Comment by sukanta-27** [comment]
@whipermr5 

>I think we can. This issue/PR can be about fixing the places that misuse these methods.

Okay. I will start my work on that. And will also modify the documentation for `isOpened()` and `isWaitingToOpen()` based on their current implementation.

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java [hunk 2]
```diff
@@ -227,7 +227,14 @@ public boolean isValid() {
...
     }
 
+    /**
+     * Returns {@code true} if the session is closed after the given hours,
+     * {@code false} otherwise, or if the startTime or endTime is {@code null}.
+     */
     public boolean isClosedAfter(int hours) {
+        if (endTime == null || startTime == null) {
+            return false;
+        }
```
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: I'm thinking we can use `Assumption.assertNotNull` instead to ensure that these methods are not called for private sessions. If we go this way, we don't need to talk about the null case in the javadoc.

Comment (sukanta-27) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: Okay.

Comment (sukanta-27) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: Should I use `Assumption.assertNotNull` for all methods? Like `isClosed()`, `isInGracePeriod`, `isOpened()` as these methods are not meant for private sessions as well?

Comment (whipermr5) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: Ideally yes, but need to check all their usages first.

**Comment by sukanta-27** [comment]
@whipermr5 Instead of dealing with every method separately, why don't we prevent users from creating `Private` sessions with `startTime == null` or `endTime == null`, is it a necessary feature?

**Comment by whipermr5** [comment]
Actually I think it'd be good to pause here and check how well the private session feature is used before deciding on the next course of action. If it is hardly used, removing it altogether might be less effort. @damithc can help to run a query on the live server for `FeedbackSession` entities with `feedbackSessionType` set to `PRIVATE`?

**Comment by sukanta-27** [comment]
@whipermr5 @wkurniawan07 
That would be a good idea. As, @damithc sir, also recommended that in https://github.com/TEAMMATES/teammates/pull/8408#issuecomment-363769833.

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java (hunk 2 above)
Comment (LiHaoTan) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: Sorry to interrupt, but can we use `Preconditions.checkNotNull` if we are doing this? I think we should prefer to stop using the `Assumption` class.

Comment (sukanta-27) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: @whipermr5 @LiHaoTan  If we are going this way, then I have no problem in doing so. But just out of curiosity, what is the difference between the two?

**Comment by damithc** [comment]
We only have about 30 private sessions (out of 28k sessions).  Given that low usage, I don't mind removing that feature. Just need to have a plan for dealing with existing private sessions. For example, we can inform the user that they will become visible to other instructors in the course after a certain date. That means we need to run a migration script _before_ deploying the version that doesn't support private sessions.

**Comment by sukanta-27** [comment]
@damithc @whipermr5 In that case, should I close this issue/PR and create a new issue for the removal of `Private` sessions?

**Comment by damithc** [comment]
> @damithc @whipermr5 In that case, should I close this issue/PR and create a new issue for the removal of Private sessions?

Go ahead.

**Comment by sukanta-27** [comment]
Closing because of #8447

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java (hunk 2 above)
Comment (LiHaoTan) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: @sukanta-27 just for your information.

The assumption class is copied from Junit so the semantics is not the same as what we want.

Java already has `Objects#requireNonNull` which throws `NullPointerException`. Whether we agree with that is the exception to be thrown that is the Java conventions and it has become so common that we need good reason to throw another exception 

The method also intentionally allow chaining.

Using `Preconditions` is just for consistency.

**Pull request closed without merging**